File: pbmx_ffi/java/io/rmf/pbmx/payloads/PublishSharesPayload.java
package io.rmf.pbmx.payloads;

import io.rmf.pbmx.BlockBuilder;
import io.rmf.pbmx.Id;
import io.rmf.pbmx.Share;
import java.util.Collection;
import java.util.ArrayList;

public final class PublishSharesPayload extends Payload {
    public PublishSharesPayload(Id stack, Collection<Share> shares, Collection<Share.Proof> proofs) {
        this.stack = stack;
        this.shares = new ArrayList(shares);
        this.proofs = new ArrayList(proofs);
    }

    @Override
    public void addTo(BlockBuilder builder) {
        builder.publishShares(stack, shares, proofs);
    }

    private final Id stack;
    private final Collection<Share> shares;
    private final Collection<Share.Proof> proofs;
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/payloads/PublishSharesPayload.kt
package io.rmf.pbmx.payloads

import com.sun.jna.Pointer
import com.sun.jna.ptr.LongByReference
import io.rmf.pbmx.Id
import io.rmf.pbmx.Share
import io.rmf.pbmx.ffi.FFI
import io.rmf.pbmx.ffi.RawFingerprint
import io.rmf.pbmx.ffi.RawShare
import io.rmf.pbmx.jnaArrayOf

class PublishSharesPayload(handle: Pointer) : Payload(handle) {
    val stack: Id
    val shares: Collection<Share>
    val proofs: Collection<Share.Proof>

    init {
        val length = LongByReference()
        var r = FFI.pbmx_unwrap_publish_shares(this.handle, null, null, length, null)
        assert(r == 0)

        val outId = RawFingerprint()
        val outShares = jnaArrayOf(RawShare(), length.value.toInt())
        val outProofs = Array(length.value.toInt()) { Pointer.NULL }
        r = FFI.pbmx_unwrap_publish_shares(this.handle, outId, outShares, length, outProofs)
        assert(r != 0)
        this.stack = Id(outId)
        this.shares = outShares.map { Share(it) }.toList()
        this.proofs = outProofs.map { Share.Proof(it) }.toList()
    }
}
